File: app/game_commands/heal_unit.rb
class HealUnit < GameCommand
  def initialize(game, user, params)
    super(game, user, params)
    @x, @y = *@params.values_at(:x, :y)
  end

  def execute
    require_current_player
    require_running_game

    base = require_friendly_base(@x, @y)
    unit = require_current_player_unit(@x, @y)

    raise CommandError.new("Unit cannot be repaired there.") unless base.can_build_unit_type?(unit.unit_type)
    raise CommandError.new("Unit has not taken damage.") unless unit.health < 10
    raise CommandError.new("Cannot heal after attacking.") unless unit.attacked == false
    raise CommandError.new("Unit has already healed this turn.") unless !unit.healed

    modify(unit, :health, unit.health + ((10 - unit.health) / 2.0).ceil)
    modify(unit, :movement_points, 0)
    modify(unit, :attacks, unit.attack_phases)
    modify(unit, :healed, true)
  end
end

File: public/javascripts/ec/game/game_commands/heal_unit.js
var HealUnit = new JS.Class(GameCommand, {
  initialize: function(game, user, params) {
    this.callSuper();
    this.x = params.x;
    this.y = params.y;
  },

  class_name: function() {
    return 'HealUnit';
  },

  execute: function() {
    var unit = this.game.unit_at(this.x, this.y);

    this.modify(unit, 'health', unit.health + Math.ceil((10 - unit.health) / 2))
    this.modify(unit, 'movement_points', 0);
    this.modify(unit, 'attacks', unit.def.attack_phases);
    this.modify(unit, 'healed', true);
  },

  can_undo: function() {
    return true;
  }
});
